blog/2024-12-06-deadlock-gadget-lfx-mentorship.md

```diff
@@ -0,0 +1,144 @@
+Consider this blog a report on my work during the mentorship.
+
+<!-- truncate -->
+
+## What have I been up to?
+
+My project involved developing a [new gadget](https://github.com/inspektor-gadget/inspektor-gadget/issues/3194) for detecting potential deadlocks in multi-threaded user space applications, using eBPF and WASM.
+
+Debugging and preventing deadlocks in modern, complex codebases can be challenging, and this gadget simplifies the process by identifying risks early, promoting safer and more efficient resource management.
+
+## Who are Inspektors and what are they inspekting?
+
+If you’re new here, welcome! [Inspektor Gadget](https://www.inspektor-gadget.io/) is a systems inspection framework leveraging eBPF-powered gadgets encapsulated into deployable [OCI images](https://opencontainers.org/) to provide powerful observability and debugging capabilities across environments, from container runtimes to multi-node Kubernetes clusters, with everything you need to collect, filter, format and export valuable systems data. Check out some of these gadgets [here](https://www.inspektor-gadget.io/docs/latest/gadgets/). Moving on...
+
+## What’s a deadlock?
+
+A [deadlock](https://en.wikipedia.org/wiki/Deadlock_(computer_science)) in a system occurs when a process becomes stuck, unable to proceed because each thread (sub-process) is waiting for a resource that another thread is holding. Since none of the threads can continue, the system remains in a state of permanent lock-up unless some external action is taken.
+
+Let’s take an example of a process with three mutexes and three threads.
+A mutex (_mutual exclusion_) is an object used to control access to a mutable resource (variables, files, etc.) and can be locked by a thread to indicate that the resource is currently in use by that thread. If any other thread wants access to that resource, all they can do is wait for the previous thread to unlock that mutex. This ensures data consistency as no two threads modify the resource simultaneously.
+
+Now let’s say we reach a state as depicted in the image.
+
+![A diagram depicting a deadlock](/media/2024-12-06-deadlock-depiction.svg)
+
+Thread 1, currently using mutex 1, requests access to lock mutex 2, but has to wait as mutex 2 is being used by thread 2. Now thread 2 requests access to mutex 3, but also has to wait as mutex 3 is being used by thread 3. Finally, thread 3 requests access to mutex 1, but has to wait as, in case you forgot, mutex 1 is being used by thread 1.
+We have come full circle ending up with the threads being stuck in an infinite wait cycle due to mutex lock order inversions, thus hanging the process.
+
+:::tip
+
+The diagram above represents a directed graph. Keep reading to learn how we use it to detect these deadlocks.
+
+:::
```

Comment:
This syntax is with the presumption that we are using docusaurus ([ref](https://docusaurus.io/docs/next/markdown-features/admonitions)).